assimilation_code/modules/assimilation/assim_tools_mod.f90

```diff
@@ -635,9 +640,11 @@ subroutine filter_assim(ens_handle, obs_ens_handle, obs_seq, keys,           &
          orig_obs_prior_var  = obs_ens_handle%copies(OBS_PRIOR_VAR_START:  &
             OBS_PRIOR_VAR_END, owners_index)
 
-         ! If QC is okay, convert this observation ensemble from probit to regular space
-         call transform_from_probit(ens_size, obs_ens_handle%copies(1:ens_size, owners_index) , &
-            obs_dist_params(owners_index), obs_ens_handle%copies(1:ens_size, owners_index))
+         ! Convert this observation ensemble from probit back to regular space if to_probit was successful
+         if(obs_probit_trans_ok(owners_index)) then 
+            call transform_from_probit(ens_size, obs_ens_handle%copies(1:ens_size, owners_index) , &
+               obs_dist_params(owners_index), obs_ens_handle%copies(1:ens_size, owners_index))
```

Comment:
this is still dangerous since you have the same input for intent(in) and intent(out), but the original code does that so yippee ki yay I guess.

https://github.com/NCAR/DART/blob/f0e7f99138df52600ef36caf0af1187ba81da83c/assimilation_code/modules/assimilation/probit_transform_mod.f90#L596-L601

Reply:
since you return the original if the transform fails, might be a good time to change this to inout

Reply:
i see that calling a routine with the same array with both an intent(in) and intent(out) goes several levels deep in this code so it would be pretty disruptive to change it.  but in the future i'd avoid passing in the same array multiple times with different intents.  it's leaving a trap for future code maintainers because someone looking only at the called subroutines could decide to set the outgoing state_ens(:) to 0 or MISSING_R8 as an initial value for debugging, which would result in the probit_ens(:) mysteriously getting overwritten without any protection from the intent(in).  you're requiring some behavior in the called subroutine because of how the calling code is written.

Reply:
I don't think it is standard conforming to alias in this way so not just people changing the code, its undefined what the compiler will do.